### PR TITLE
Master hr expense usability fda

### PR DESCRIPTION
**Pull request by fda-odoo**
https://www.odoo.com/web#id=30933&view_type=form&model=project.task&action=333&active_id=965&menu_id=4720

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/hr_expense/models/hr_expense.py [hunk 1]
```diff
@@ -556,7 +584,10 @@ def refuse_sheet(self, reason):
 
     @api.multi
     def approve_expense_sheets(self):
-        self.write({'state': 'approve', 'responsible_id': self.env.user.id})
+        responsible = self.responsible_id.id
```
Comment (fda-odoo) [inline addons/hr_expense/models/hr_expense.py]: responsible =  self.responsible_id.id or self.env.user.id

File: addons/hr_expense/views/hr_expense_views.xml [hunk 2]
```diff
@@ -87,6 +88,38 @@
...
+            <field name="name">hr.expense.form</field>
+            <field name="model">hr.expense</field>
+            <field name="inherit_id" ref="hr_expense.hr_expense_form_view_without_header"/>
+            <field name="mode">primary</field>
+            <field name="arch" type="xml">
+                <xpath expr="/form/*" position="before">
+                    <header>
+                        <!-- Buttons when there is no attachment with the expense -->
+                        <button name="submit_expenses" string="Submit to Manager" type="object" attrs="{'invisible': ['|', ('attachment_number', '=', 0), ('state', 'not in', 'draft')]}" class="oe_highlight o_expense_submit"/>
+                        <button name="upload_document" string="Upload a document" type="object" attrs="{'invisible': ['|', ('attachment_number', '=', 0), ('state', 'not in', 'draft')]}" class="o_upload_document"/>
```
Comment (bst-odoo) [inline addons/hr_expense/views/hr_expense_views.xml]: The upload document feature has been done here : https://www.odoo.com/web#id=36712&view_type=form&model=project.task&action=333&active_id=131&menu_id=4720
So it has to be removed from this task. Sorry there was some miscommunication about it in this task.

File: addons/hr_expense/models/hr.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# -*- coding: utf-8 -*-
```
Comment (rim-odoo) [inline addons/hr_expense/models/hr.py]: you could name this file `hr_employee`

File: addons/hr_expense/models/hr.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
```
Comment (rim-odoo) [inline addons/hr_expense/models/hr.py]: `api` seems unused :)

File: addons/hr_expense/models/hr_expense.py [hunk 5]
```diff
@@ -428,8 +451,8 @@ class HrExpenseSheet(models.Model):
     help='Expense Report State')
     employee_id = fields.Many2one('hr.employee', string="Employee", required=True, readonly=True, states={'submit': [('readonly', False)]}, default=lambda self: self.env['hr.employee'].search([('user_id', '=', self.env.uid)], limit=1))
     address_id = fields.Many2one('res.partner', string="Employee Home Address")
-    payment_mode = fields.Selection([("own_account", "Employee (to reimburse)"), ("company_account", "Company")], related='expense_line_ids.payment_mode', default='own_account', readonly=True, string="Payment By")
-    responsible_id = fields.Many2one('res.users', 'Validation By', readonly=True, copy=False, states={'submit': [('readonly', False)]})
+    payment_mode = fields.Selection([("own_account", "Employee (to reimburse)"), ("company_account", "Company")], related='expense_line_ids.payment_mode', default='own_account', readonly=True, string="Paid By")
+    user_id = fields.Many2one('res.users', 'Manager', readonly=True, copy=False, states={'submit': [('readonly', False)]}, track_visibility='onchange')
```
Comment (rim-odoo) [inline addons/hr_expense/models/hr_expense.py]: if you rename a field (`responsible_id` as `user_id`), you should use the attribute `oldname=` in order to ease migration ;)

File: addons/hr_expense/models/hr_expense.py [hunk 6]
```diff
@@ -519,11 +545,12 @@ def _create_set_followers(self, values):
         MailFollowers = self.env['mail.followers']
         for partner in users.mapped('partner_id'):
             values['message_follower_ids'] += MailFollowers._add_follower_command(self._name, [], {partner.id: None}, {})[0]
-
+
```
Comment (rim-odoo) [inline addons/hr_expense/models/hr_expense.py]: just a detail: don't let your linter mess the diff!

File: addons/hr_expense/views/hr_expense_views.xml [hunk 7]
```diff
@@ -760,11 +773,12 @@ if records:
...
-        <menuitem id="menu_hr_expense_sheet_all_all" name="Expenses Reports Analysis" sequence="4" parent="menu_hr_expense_reports"
-            action="action_hr_expense_sheet_all_all" groups="hr_expense.group_hr_expense_manager"/>
 
         <menuitem id="menu_hr_expense_configuration" name="Configuration" parent="menu_hr_expense_root"
             sequence="100"/>
         <menuitem id="menu_hr_product" name="Expense Products" parent="menu_hr_expense_configuration"
             action="hr_expense_product" groups="hr_expense.group_hr_expense_manager" sequence="10"/>
+
+
+
```
Comment (rim-odoo) [inline addons/hr_expense/views/hr_expense_views.xml]: great diff :D

File: addons/hr_expense/models/hr_expense.py (hunk 5 above)
Comment (tde-banana-odoo) [reply addons/hr_expense/models/hr_expense.py]: +1

File: addons/hr_expense/models/hr_expense.py [hunk 8]
```diff
@@ -552,7 +583,8 @@ def refuse_sheet(self, reason):
     def approve_expense_sheets(self):
         if not self.user_has_groups('hr_expense.group_hr_expense_user'):
             raise UserError(_("Only HR Officers can approve expenses"))
-        self.write({'state': 'approve', 'responsible_id': self.env.user.id})
+        responsible = self.user_id.id or self.env.user.id
```
Comment (tde-banana-odoo) [inline addons/hr_expense/models/hr_expense.py]: Actually you should override responsible if not already set. What is the intended behavior ?

File: addons/hr_expense/views/hr_expense_views.xml [hunk 9]
```diff
@@ -1,6 +1,19 @@
 <?xml version="1.0" encoding="utf-8"?>
 <odoo>
 
+        <record id="hr_employee_view_form_inherit_expense" model="ir.ui.view">
+            <field name="name">hr.employee.form.expense</field>
```
Comment (tde-banana-odoo) [inline addons/hr_expense/views/hr_expense_views.xml]: SmallNote: hr.employee.view.form

File: addons/hr_expense/views/hr_expense_views.xml [hunk 10]
```diff
@@ -21,26 +34,13 @@
             </field>
         </record>
 
-        <record id="hr_expense_form_view" model="ir.ui.view">
+        <record id="hr_expense_form_view_without_header" model="ir.ui.view">
```
Comment (tde-banana-odoo) [inline addons/hr_expense/views/hr_expense_views.xml]: SmallNote: hr_expense_view_form_ ...

File: addons/hr_expense/views/hr_expense_views.xml [hunk 11]
```diff
@@ -87,6 +88,33 @@
             </field>
         </record>
 
+        <record id="hr_expense_form_view" model="ir.ui.view">
```
Comment (tde-banana-odoo) [inline addons/hr_expense/views/hr_expense_views.xml]: SmallNote: hr_expense_view_form

File: addons/hr_expense/views/hr_expense_views.xml [hunk 12]
```diff
@@ -87,6 +88,33 @@
...
+            <field name="name">hr.expense.form</field>
```
Comment (tde-banana-odoo) [inline addons/hr_expense/views/hr_expense_views.xml]: SmallNote: hr.expense.view.form

File: addons/hr_expense/views/hr_expense_views.xml [hunk 13]
```diff
@@ -87,6 +88,33 @@
...
+            <field name="model">hr.expense</field>
+            <field name="inherit_id" ref="hr_expense.hr_expense_form_view_without_header"/>
+            <field name="mode">primary</field>
+            <field name="arch" type="xml">
+                <xpath expr="/form/*" position="before">
+                    <header>
+                        <button name="submit_expenses" states="draft" string="Submit to Manager" type="object" class="oe_highlight o_expense_submit"/>
```
Comment (tde-banana-odoo) [inline addons/hr_expense/views/hr_expense_views.xml]: What is the purpose of the o_expense_submit class ?

File: addons/hr_expense/views/hr_expense_views.xml [hunk 14]
```diff
@@ -385,19 +413,21 @@
             <field name="arch" type="xml">
                 <form string="Expense Reports" class="o_expense_sheet">
                  <header>
-                    <button name="approve_expense_sheets" states="submit" string="Approve" type="object" groups="hr_expense.group_hr_expense_user" class="oe_highlight o_expense_sheet_approve"/>
+                    <field name="is_user_the_expense_responsible" invisible="1"/>
+                    <button name="approve_expense_sheets" string="Approve" type="object" attrs="{'invisible': ['|', ('state', '!=', 'submit'), ('is_user_the_expense_responsible', '=',False)]}" class="oe_highlight o_expense_sheet_approve"/>
```
Comment (tde-banana-odoo) [inline addons/hr_expense/views/hr_expense_views.xml]: Generally we avoid hiding buttons for things like "I am not the responsible". It should be based on access rights as on all main applications. What do you think about that change ?

File: addons/hr_expense/views/hr_expense_views.xml [hunk 15]
```diff
@@ -385,19 +413,21 @@
...
                     <button name="action_sheet_move_create" states="approve" string="Post Journal Entries" type="object" groups="account.group_account_manager" class="oe_highlight o_expense_sheet_post"/>
                     <button name="%(hr_expense.hr_expense_sheet_register_payment_wizard_action)d" type="action" string="Register Payment" class="oe_highlight o_expense_sheet_pay" attrs="{'invisible': [('state', '!=', 'post')]}" context="{'default_amount': total_amount, 'partner_id': address_id}" groups="account.group_account_manager"/>
                     <button name="reset_expense_sheets" states="cancel" string="Resubmit" type="object"/>
-                    <button name="%(hr_expense.hr_expense_refuse_wizard_action)d" states="submit,approve" context="{'hr_expense_refuse_model':'hr.expense.sheet'}" string="Refuse" type="action" groups="hr_expense.group_hr_expense_user" />
+                    <button name="%(hr_expense.hr_expense_refuse_wizard_action)d" context="{'hr_expense_refuse_model':'hr.expense.sheet'}" string="Refuse" type="action" attrs="{'invisible': ['|', ('state', 'not in', ('submit','approve')), ('is_user_the_expense_responsible', '=',False)]}" />
```
Comment (tde-banana-odoo) [inline addons/hr_expense/views/hr_expense_views.xml]: Generally we avoid hiding buttons for things like "I am not the responsible". It should be based on access rights as on all main applications. What do you think about that change ?

File: addons/hr_expense/models/hr.py [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+from odoo.tools import float_round
```
Comment (tde-banana-odoo) [inline addons/hr_expense/models/hr.py]: Imports should be cleaned

File: addons/hr_expense/views/hr_expense_views.xml (hunk 14 above)
Comment (tde-banana-odoo) [inline addons/hr_expense/views/hr_expense_views.xml]: If really necessary can't you use a domain like 'user_id', '=', uid ?

File: addons/hr_expense/views/hr_expense_views.xml [hunk 17]
```diff
@@ -50,6 +50,7 @@
                     <group>
                         <group>
                             <field name="product_id" context="{'default_can_be_expensed': 1}"/>
+                            <field name="state" widget="statusbar" statusbar_visible="draft,reported,approved,done,refused" invisible="1"/>
```
Comment (jem-odoo) [inline addons/hr_expense/views/hr_expense_views.xml]: Why invisible, if your other view add the statusbar ?

File: addons/hr_expense/models/hr_expense.py [hunk 18]
```diff
@@ -137,7 +140,9 @@ def get_empty_list_help(self, help_message):
                     _('Click to add a new expense,'),
                     _('or send receipts by email to %s.') % (link,),
                     help_message)
-        return super(HrExpense, self).get_empty_list_help(help_message)
+        return '<p class="oe_view_nocontent_create">%s</p>%s' % (
+            _('Click to add a new expense.'),
+            help_message)
```
Comment (jem-odoo) [inline addons/hr_expense/models/hr_expense.py]: Why the `super`is not called anymore. I don't understand what is this change :s

**Pull request merged**